crates/storage/src/collections/slice/mod.rs

```diff
@@ -0,0 +1,384 @@
+{
+    /// Returns the number of elements in the slice.
+    pub fn len(&self) -> u32 {
+        self.range.end - self.range.start
+    }
+
+    /// Returns true if the slice has a length of 0.
+    pub fn is_empty(&self) -> bool {
+        self.range.end == self.range.start
+    }
+
+    /// Returns the first element of the slice, or None if it is empty.
+    pub fn first(&self) -> Option<&T::Item> {
+        self.get(0)
+    }
+
+    /// Returns the first element of the slice, or None if it is empty.
+    pub fn last(&self) -> Option<&T::Item> {
+        self.get(self.len())
+    }
+
+    /// Returns the first and all the rest of the elements of the slice, or None if it is empty.
+    pub fn split_first(&self) -> Option<(&T::Item, Slice<&T>)> {
+        let first = self.first()?;
+        Some((
+            first,
+            Slice::new(self.range.start + 1..self.range.end, &self.backing_storage),
+        ))
+    }
+
+    /// Returns the last and all the rest of the elements of the slice, or None if it is empty.
+    pub fn split_last(&self) -> Option<(&T::Item, Slice<&T>)> {
+        let first = self.last()?;
+        Some((
+            first,
+            Slice::new(self.range.start..self.range.end - 1, &self.backing_storage),
+        ))
+    }
+
+    /// Creates a new `Slice`.
+    pub fn new(range: Range<u32>, backing_storage: T) -> Slice<T> {
+        Slice {
+            range,
+            backing_storage,
+        }
+    }
+
+    /// Returns a reference to an element or None if out of bounds.
+    pub fn get(&self, index: u32) -> Option<&T::Item> {
+        self.backing_storage.get(index + self.range.start)
+    }
+
+    /// Returns an iterator over the slice.
+    #[inline]
+    pub fn iter(&self) -> Iter<T> {
+        Iter {
+            index: 0,
+            range: self.range.clone(),
+            backing_storage: &self.backing_storage,
+        }
+    }
+
+    /// Divides one slice into two at an index.
+    ///
+    /// The first will contain all indices from [0, mid) (excluding the index mid itself) and the
+    /// second will contain all indices from [mid, len) (excluding the index len itself).
+    #[inline]
+    pub fn split_at<'a>(&'a self, mid: u32) -> (Slice<&T>, Slice<&T>)
+    where
+        &'a T: ContiguousStorage,
+    {
+        assert!(mid <= self.len());
+        (
+            Slice::new(0..mid, &self.backing_storage),
+            Slice::new(mid..self.len(), &self.backing_storage),
+        )
+    }
+}
+
+/// A view into a storage `Vec`.
+#[derive(Clone, Debug)]
+#[must_use = "slices are views into underlying storage and do nothing unless consumed"]
+pub struct SliceMut<T> {
+    /// The start and end indices inside the `index_map`. Indexing the slice using `n` means that we
+    /// access `n + range.start`.
+    range: Range<u32>,
+    backing_storage: T,
+}
+
+impl<T> SliceMut<T>
+where
+    T: ContiguousStorage,
+{
+    /// Returns the number of elements in the slice.
+    #[inline]
```

Comment:
Why'd you choose to `inline` these methods, but not the ones for `Slice` (and vice versa)?

Reply:
To be honest, I forgot to inline these. I mainly inline based on `core`'s implementation, as that seems reasonable.